### PR TITLE
Extract ThrottledTaskRunner

### PR DESCRIPTION
**Pull request by DaveCTurner**
Generalizes `PrioritizedThrottledTaskRunner` slightly:

- The throttling behaviour is also useful for tasks which do not complete synchronously. The new `ThrottledTaskRunner` passes a `Releasable` to each task, which until released will prevent spawning further tasks.

- The only part that needs the tasks to be `Comparable<>` is the queue. Letting the caller specify the queue means that we can also use the throttling without the prioritisation.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
I particularly would like to use this generalisation in https://github.com/elastic/elasticsearch/pull/92373, although I think we will find other uses too.

File: server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java [hunk 1]
```diff
@@ -245,23 +247,16 @@ private int getRandomPriority() {
...
         assertThat(taskRunner.runningTasks(), equalTo(0));
     }
 
-    private static void awaitBarrier(CyclicBarrier barrier) {
-        try {
-            barrier.await(10, TimeUnit.SECONDS);
-        } catch (Exception e) {
-            throw new AssertionError("unexpected", e);
-        }
-    }
```
Comment (pxsalehi) [inline server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: Aside from getting rid of this, is there any other advantage in introducing `TestBarrier`?

File: server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    private static final ThreadContext threadContext = new ThreadContext(Settings.EMPTY);
+
+    private ExecutorService executor;
+    private int maxThreads;
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        maxThreads = between(1, 10);
+        executor = EsExecutors.newScaling("test", 1, maxThreads, 0, TimeUnit.MILLISECONDS, false, threadFactory, threadContext);
```
Comment (pxsalehi) [inline server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java:38]: Does this also need to change like https://github.com/elastic/elasticsearch/pull/93446?

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.core.Releasable;
+
+import java.util.concurrent.Executor;
+
+public class ThrottledTaskRunner extends AbstractThrottledTaskRunner<ActionListener<Releasable>> {
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java:16]: Is this used anywhere? If not, could we introduce it, when it is needed?

**Review by pxsalehi — COMMENTED** [review]
Thanks David. Added some questions. Aside from using the`ActionListener<Releasable>` there doesn't seem to be a lot of changes in terms of concurrency to the task runner itself? Is that correct?

File: server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: No, it's just about those pointless checked exceptions.

File: server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/AbstractThrottledTaskRunnerTests.java:38]: Yes.

File: server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java:16]: I will be using it in https://github.com/elastic/elasticsearch/pull/92373 which is already quite a big change so I'd like to have prerequisites like this reviewed separately. It also came up in a discussion about search execution so I wanted to have the code somewhere visible. I'm happy to delay merging it until #92373 is closer to ready tho.

File: server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: (I can do this separately too, there's lots of other spots where this is useful)

Comment (pxsalehi) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: TBH, I find the class unnecessary. A utility function which takes a timeout parameter will do as well. But I'm fine with leaving it, if you'd like.

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: Ok, I'll do this in a separate PR

Comment (pxsalehi) [reply server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:266]: If you find it useful, it's perfectly fine here. I was just trying to understand why we need it at all. 😄

**Review by pxsalehi — APPROVED** [review]

**Comment by pxsalehi** [comment]
> Does this also need to change like https://github.com/elastic/elasticsearch/pull/93446?

@DaveCTurner There were two `EsExecutors.newScaling` executors here in the test which do not seem to have the "fix" that you had in https://github.com/elastic/elasticsearch/pull/93446. Is it not needed here to prevent flaky tests?

**Comment by DaveCTurner** [comment]
Ah sorry you're right, I forgot about this new test suite. I opened https://github.com/elastic/elasticsearch/pull/93505.

**Pull request merged**